KNI [Kubernetes Networking Interface] Initial Draft KEP (#4477)
kubernetes/enhancements: keps/sig-network/4410-k8s-network-interface/README.md

# KEP-4410: Kubernetes Networking reImagined

> **NOTE**: for the initial PR we've removed a lot of the templated text and
> aimed to keep this first iteration small and easier to consume. We are only
> focusing on the "What" and "Why" (e.g. motivation, goals, user stories) for
> this iteration so that we can build consensus on those first before we add
> any of the "How".

<!-- toc -->
- [Summary](#summary)
- [Motivation](#motivation)
  - [Goals](#goals)
  - [Non-Goals](#non-goals)
- [Proposal](#proposal)
  - [User Stories (Optional)](#user-stories-optional)
    - [Story 1](#story-1)
    - [Story 2](#story-2)
<!-- /toc -->

## Summary

> This proposal is to design and implement the KNI [Kubernetes Networking Interface] or better known as Kubernetes Networking reImagined. KNI will create a Network resource and provide an API that will provide network status, availability, how to attach a pod to a network, detach the pod from the network and update a pods network.  

Reviewer: I don't think it's "better" known as anything, given that it mostly does not exist yet.

The Summary and Motivation should not assume that the reader is already familiar with the idea.

Author: It is funny you mention this because reImagined is how it got recirculated back to me and I just ran with it. However, 100% the summary/motivation should be written in a way that a nontechnical reader should understand. However, that is a high bar to hit as the current is very difficult and through the years, I have found that small numbers of people can actually articulate the current accurately.

Reviewer 2: ```suggestion
This proposal is to design and implement a KNI [Kubernetes Networking Interface]. KNI will create a Network resource and provide an API that will provide network status, availability, how to attach a pod to a network, detach the pod from the network and update a pods network.  
```

Reviewer 3: This is my first time trying to get familiar with the KNI project, so started reading this, Apologies if this sounds like a total newbie question, perhaps I must read other material first before coming to this KEP and if so please point me the right way:
```
provide an API that will provide network status, availability, how to attach a pod to a network, detach the pod from the network and update a pods network
```
`network status and availability` means the cluster networking status? node networking status? pod networking status?

the second part refers to pod aspects:
`how to attach a pod to a network, detach the pod from the network and update a pods network` => this is like providing what the CNI spec does but through an API right?

## Motivation

Kubernetes networking has traditionally been challenging to understand for users
interacting with the Kubernetes API, and there has been considerable flexibility
in how Container Network Interfaces (CNIs) set up networking within clusters.
This has resulted in a scenario where things like pod networking (including pod
to pod networking) is opaque to users, with different implementations taking

> markedly different approaches. This fragmentation has spread networking across

Reviewer: So... KNI will reduce the flexibility in how pod networking can be set up, so that all implementations will have to take similar approaches?

Author: Reading this now and do see that this implies that we might be reducing flexibility which is the exact opposite of what we are trying to achieve.

Reviewer 3: also the same with pod2pod networking is opaque to users aspect => Isn't that mostly how it should be as in end users only want the workloads to communicate correctly don't like 90% of end users not care about the "how" ? so opaque is not necessarily bad..right? :)

Reviewer 3: and +1 to your previous comment:
```
reducing flexibility which is the exact opposite of what we are trying to achieve.
```
would love to see how we plan to achieve this, i.e bring a common API used by all but adoptable enough by different implementations..

all layers of the stack which include k8s components like kube-proxy, netpol agents,
container runtime with CNI plugins and low level runtimes like kata and issues

> with the API have negatively impacted adoption in sectors such as telecommunications.

Reviewer: "Issues"? Just randomly saying that "issues" have negatively impacted adoption gives no information on what's wrong or how KNI will fix it.

Reviewer 4: I'd like to see a description of these issues that cause "negatively impacted adoption" , everytime I talk with someone of this problem the discussion boils down to they want to use the Pods as VMs and doesn't work, so it is not the network per se, is that kubernetes does not work as they expect

Author: I think I should merge my suggested motivation as the current motivation might be vague and we are spending cycles on something that is going away.

Our goal is to transform Kubernetes networking by making networks and their components
actual resources within the Kubernetes API. This will allow for the development

> of shared functionalities and their integration into the API. We anticipate that

Reviewer: "shared functionalities"?

> this new approach will enhance support for areas that are currently struggling,

Reviewer: what areas are currently struggling?

> facilitate the development and promotion of common features, and better define

Reviewer: what sort of features, and why can't they be developed now?

> and accommodate advanced functionalities and potential areas for expansion.

Reviewer: You started out the section implying that the problem you wanted to solve was that Kubernetes networking was hard for users to understand. But then you end by talking about how you will help network plugin developers implement new features... Are these new features going to make Kubernetes networking easier for users to understand? If so, how?

### Goals

1. Design a cool looking t-shirt
2. Provide Kubernetes APIs for the creation, configuration and management of networks (e.g. `Pod` networks)
3. Provide documentation, examples, troubleshooting and FAQ's for KNI.

> 4. Establish feature parity with current CNI [ADD, DEL]

Reviewer: So as an example of what is wrong with this KEP, you say here that you're going to provide feature parity with CNI, but _you never said that this is a replacement for CNI_.

Author: Should we say this is a replacement for the CNI? In the POC I leverage the CNI as a means to prove out a migration strategy. It becomes an implementation detail. Should we rephrase this as we need a means to attach and detach a network without referring to the existing CNI ADD/DEL?

Reviewer: You should say what is _is_. If it's not a replacement for CNI then I'm already misunderstanding it!

> 5. Handle support levels like Gateway API (e.g. "core" and "extended")

Reviewer: support levels of what?

6. Handle implementation-specific use cases through extension points
7. Decouple the Pod and Node Network setup

> 8. Simplify/enable triggering garbage collection to ensure no resources are left behind

Reviewer: garbage collection of what?

Author: Should we include specific examples of ebpf programs, linux network artifacts such as linux bridges?

Reviewer: my point was more that you are giving goals like "we will garbage collect things" without having mentioned the existence of anything that will need to be garbage-collected.

> 9. Provide the ability to identify the IP address family without parsing the value (such as a field)

Reviewer: IP address family of what?

10. Provide as much backwards-compatibility with CNI as is feasible
11. Guarantee the network is setup and in a healthy state before containers are started (ephemeral, init, regular)
12. If feasible, provide API awareness of Pod network namespaces (e.g. interface names)
13. Provide support for Kata and other virtualized runtimes

> 14. Provide a reference implementation

Reviewer: A reference implementation of what?

# KEP-4410: Kubernetes Networking reImagined

> **NOTE**: for the initial PR we've removed a lot of the templated text and
> aimed to keep this first iteration small and easier to consume. We are only
> focusing on the "What" and "Why" (e.g. motivation, goals, user stories) for
> this iteration so that we can build consensus on those first before we add
> any of the "How".

<!-- toc -->
- [Summary](#summary)
- [Motivation](#motivation)
  - [Goals](#goals)
  - [Non-Goals](#non-goals)
- [Proposal](#proposal)
  - [User Stories (Optional)](#user-stories-optional)
    - [Story 1](#story-1)
    - [Story 2](#story-2)
<!-- /toc -->

## Summary

> This proposal is to design and implement the KNI [Kubernetes Networking Interface] or better known as Kubernetes Networking reImagined. KNI will create a Network resource and provide an API that will provide network status, availability, how to attach a pod to a network, detach the pod from the network and update a pods network.  

Author: ```suggestion
KNI or Kubernetes Networking Interface, is an effort that is to take a second look at Kubernetes networking and evaluate what are the painpoints, and what can we improve. At its core, KNI will be a foundational Kubernetes specific networking API that will provide flexibility and estensibility to the solve basic and the most advanced and opinionated networking use cases.  
 
```

## Motivation

Kubernetes networking has traditionally been challenging to understand for users
interacting with the Kubernetes API, and there has been considerable flexibility
in how Container Network Interfaces (CNIs) set up networking within clusters.
This has resulted in a scenario where things like pod networking (including pod
to pod networking) is opaque to users, with different implementations taking
markedly different approaches. This fragmentation has spread networking across
all layers of the stack which include k8s components like kube-proxy, netpol agents,
container runtime with CNI plugins and low level runtimes like kata and issues
with the API have negatively impacted adoption in sectors such as telecommunications.

> Our goal is to transform Kubernetes networking by making networks and their components

Reviewer 3: what does "components" here mean?

- If feasible, provide API awareness of Pod network namespaces (e.g. interface names)
- Provide a uniform approach for network setup/teardown for both virtualized (kata) and non-virtualized (runc) 
runtimes including kubevirt. This could eliminate the high and low level runtimes from the networking path
- Provide a reference implementation of the KNI network runtime
- Provide the ability to have all the dependencies packaged in the container image (no more CNI binaries in the host file system)
..- No more downloading CNI binaries via initContainers/Mounting /etc/cni/net.d or /opt/cni/bin
- Provide the ability to use native k8s resources for configuration such as a ConfigMap's instead of configuration files in host file system
- Provide an API to indicate network readiness for the node (no more files on disk)
- Eliminate the need to exec binaries and replace with gRPC
- Make troubleshooting easier by having logs accessible via kubectl logs
- Improve network pod startup time
- Provide the ability to prevent additional scheduling of pods if IPAM is out of IP addresses without evicting running pods

> 

Reviewer 5: TODO: Add goal of having the pod object available at network runtime

Author: @dougbtv I am drafting an update. So I might be able to get this. Do you have specific items you want off the Pod spec? Metadata (name, namespace, labels, annotations, ... )

Reviewer 5: Metadata nails it, thanks. At least I'm most interested in getting all you listed. Potentially someone might want more?

Reviewer 6: I'd like to request CIDRs as an available piece of metadata if possible. That would be great for legacy applications that use CIDR configurations as config values (in my case, Ceph).

Author: @BlaineEXE this sounds reasonable. I notice you are in Colorado. I am in the boulder area. Does the application need the pod cidr? You might be able to infer this via the Pod IP.

Reviewer 6: Hey! Yeah, I'm in Denver. And also a climber :)

Our storage application [Ceph](https://ceph.io) does require CIDR in certain circumstances. In "basic" operation, we don't need to provide the CIDR of the pod network.

But when we use Multus to attach pods to host devices (described more [here](https://github.com/kubernetes/enhancements/pull/4477/files#r1486407995)), we need to configure Ceph to listen on a CIDR (or list of CIDRs) so that it knows which interface(s) to use. I don't fully grok why Ceph uses CIDR instead of iface names, but it's what we have to work with.

Currently, what we have been doing is to use the JSON output of `ip addr` in a K8s Job to discover the CIDR of the network. It has been working for us, but it's a step that would be nice to eliminate if possible.

Author: Let's climb! Are you at the spot or movement? This sounds like a story for the KNI that could enhance the downward API information with additional node specific network information such as Pod CIDR?

Reviewer 4: > . I don't fully grok why Ceph uses CIDR instead of iface names, but it's what we have to work with.

we need to be careful to not absorb other projects problems, and avoid bespoke solutions

Reviewer 6: @MikeZappa87 I'm with the Spot and usually get punches for mvmt.

@aojea I agree that this project shouldn't absorb other problems. I can offer up our use case (attaching pods to raw devices, potentially with macvlan) as something that seems worthwhile for KNI to consider. 

I personally think that this part of our use case is extremely valuable applications that want to achieve maximum network performance while also benefitting from the superior scheduling offered by Kubernetes. 

I have been conceptualizing this similarly to how Kubernetes allows workflows that integrate GPUs into pod environments. Some nodes have special, high-performance hardware available to provide acceleration for certain workflows. As GPUs are workflow accelerators, so too are networks.

As far as the CIDR goes, my experience has been that every network has one, and I think it is potentially useful for that to be part of the metadata made available to pods. However, if there are deeper complexities with exposing CIDR metadata that I'm not aware of, it won't make or break my project to not have that included in KNI; we are already working around the lack of that information in the existing sig-network APIs.

Reviewer 4: >  I can offer up our use case (attaching pods to raw devices, potentially with macvlan) as something that seems worthwhile for KNI to consider.

that is exactly what we need, these experiences, and this one is something I've identified in multiple places, attach netdevices to pods, so I feel this is a strong use case ... what I also see is that these interfaces are used as "external' networks that are only relevant to the app running on the specific pod, so I don't feel that these IPs from these interfaces should be represented on the kubernetes topology ...

Reviewer 4: @BlaineEXE I'm still trying to fully understand your use case, based on your comments it seems you need to have some prior work of setting up the infrastructure and the vlans,
who configures these vlans?
are these vlans configured on all these hosts?
...

Reviewer 6: Certainly someone must configure the additional hardware. In practice, an admin must add a separate switch (or create a VLAN on an existing switch) that connects to a different interface on the host systems. So if `eth0` underpins the k8s pod network, `eth1` may be the result of the additional interface, unused by k8s itself.

Our current deployment strategy leverages Multus and NetworkAttachmentDefinitions to connect storage (Ceph) pods to `eth1`. We recommend CNI=macvlan (lower latency than bridge) with IPAM=whereabouts (ease of use), but users can practically use whatever CNI/IPAM they like.

This does work, but because Multus is such a complex feature to understand, users often seem lost trying to configure an already-complex storage system with NADs. In addition, there is developer complexity, and there are friction points -- like not being able to get a Service with a static IP on a Multus network.

Reviewer 4: >  like not being able to get a Service with a static IP on a Multus network.

what do you mean by a multus network? some entity that is connected to the additional interface?

Reviewer 4: @BlaineEXE @dougbtv I want to understand well this use case, I don't quite get what are the expectations here and what is the problem we are trying to solve., definitively kubernetes can not manage the infra, connecting switches or create vlanes, there are other projects that cover that area ... 
but this part of _Service with staticIP on Multus network_ is the one I want to understand, the docs are not clear  https://github.com/rook/rook/blob/master/design/ceph/multus-network.md, does it mean to make services available out of the cluster?

### Non-Goals

1. Any changes to the kube-scheduler 
2. Any specific implementation other than the reference implementation. However we should ensure the KNI-API is flexible enough to support

## Proposal

The proposal of this KEP is to design and implement the KNI-API and make necessary changes to the CRI-API and container runtimes. The scope should be kept to a minimum and we should target feature parity. 

### User Stories

We are constantly adding these user stories, please join the community sync to discuss. 

#### Story 1

As a cluster operator, I need the ability to determine my network(s) is ready so that my pods come up with a working network.

#### Story 2

As a cluster operator, I need the ability to determine what networks are available on my node so that upstream components can ensure the pod is scheduled on the appropriate node. 

#### Story 3

> As a Kubernetes developer, I need the ability to have extension points for pod network setup, teardown and update so that I can support future Kubernetes networking features with either reducing the changes to core kubernetes or eliminating them

Reviewer 5: So, I'm reading this as not just setup and teardown, but, within a pod's lifecycle, maybe we can expand this to say that it's not just "update" but during it's lifecycle, can that be more explicit about taking actions while the pod is running?

> I need the ability to have extension points for pod network setup, teardown and updates while pods are running

I think of this as a current limitation at having execution points on pod creation or deletion. 

Also this could be split into its own thing, because there's a kinda double thing here with both extension points and reducing or eliminating changes to core.

Reviewer 4: " kubernetes developer that wants to do network things " is not an user story, you need to define the user story that require those advanced networking features , and then we figure out what is thebest solution

> As a Kubernetes developer, I need the ability to have extension points for pod network setup, teardown and update so that I can support future Kubernetes networking features with either reducing the changes to core kubernetes or eliminating them

Reviewer 6: I have a use-case for consideration here that I can break down into 2 parts:

Part 1: Currently the [Rook](https://github.com/rook/rook) project uses Multus to allow storage pods to attach to host network devices without exposing pods on the host network namespace. The goal is to be able to keep as much network isolation for security as possible while also being able to give the Ceph storage platform access to host-local network speeds. It's not clear to me whether the high-level use case I am describing -- attaching pods to specific host devices -- has a place in KNI. I certainly hope so. (As a note here, Rook uses prefers to use macvlan to get subdevices of a host device rather than exposing the host device itself, but I can also imagine systems that want to just expose the host device itself.)

Part 2: Assuming part 1 fits into KNI's purpose, the next part relates to IP assignments. A big hangup the Rook project has with Multus is that we can't easily get static IP assignments on Multus networks. For the Pod network, it is easy to get a static IP via k8s Service, but we don't have the same ease when doing this for dedicated NICs (\*). I would propose that KNI consider whether it can create API flexibility to allow Service functionality on any network, including host-device-attached networks.

(\*): We are aware that [MultusService](https://github.com/k8snetworkplumbingwg/multus-service) exists, but it isn't at a stable enough place for us to requiring it as an add-on for users. Additionally, it requires Rook to

Reviewer 4: @MikeZappa87 this comment is proobing my point on the user story I added as example here https://github.com/kubernetes/enhancements/pull/4477#issuecomment-1944464788

Author: I think we need to rethink this user story.

Reviewer 4: >  For the Pod network, it is easy to get a static IP via k8s Service, but we don't have the same ease when doing this for dedicated NICs (*). I would propose that KNI consider whether it can create API flexibility to allow Service functionality on any network, including host-device-attached networks.

if the pod is connected through a external nic to a external network, is the external network responsibility to assign IPs, are we trying to say that kubernetes should manage these external networks ipam? what stops an admin on that network to assign a static ip?

#### Story 4

> As a tool which manages eBPF programs on a Kubernetes cluster (bpfman,

Author: I think we should keep the personal to an actually user?

# KEP-4410: Kubernetes Networking reImagined

> **NOTE**: for the initial PR we've removed a lot of the templated text and
> aimed to keep this first iteration small and easier to consume. We are only
> focusing on the "What" and "Why" (e.g. motivation, goals, user stories) for
> this iteration so that we can build consensus on those first before we add
> any of the "How".

<!-- toc -->
- [Summary](#summary)
- [Motivation](#motivation)
  - [Goals](#goals)
  - [Non-Goals](#non-goals)
- [Proposal](#proposal)
  - [User Stories (Optional)](#user-stories-optional)
    - [Story 1](#story-1)
    - [Story 2](#story-2)
<!-- /toc -->

## Summary

KNI or Kubernetes Networking Interface, is an effort to take a second look at Kubernetes networking and evaluate what are the pain points, and what can we improve. At its core, KNI will be a foundational network API specific for Kubernetes that will provide flexibility and extensibility to solve basic and the most advanced and opinionated networking use cases.  
 

> ## Motivation

Author: ```suggestion
## Motivation

Kubernetes networking is an area of complexity and multiple layers which has created several challenges and areas of improvement. These challenges include deployment of the CNI plugins, troubleshooting networking issues and development of new functionality. 

Currently networking happens in three layers of the stack, Kubernetes itself by means of kube-proxy or another controller based solution, the container runtime with network namespace creation and CNI plugins and the OCI runtime which does additional network setup for kernel isolated pods. All of this communication happens through non network specific APIs which to the reader of the code makes it hard to determine where ‘networking’ is happening. Having networking in several layers presents an issue when needing to troubleshoot issues as one needs to check several areas and some cannot be done via kubectl logs such as the CNI execution logs. This becomes more of an effort as multiple uncoordinated processes are making changes to the same resource, the network namespace of either the root or pod. The KNI aims at reducing the complexity by consolidating the networking into a single layer and having a uniform process for both namespaced and kernel isolated pods through a gRPC API. Leveraging gRPC will allow users the ability to migrate away from the current execution model that the CNI currently leverages.   

The next challenge is the deployment of the CNI plugins that provide the network setup and teardown of the pod. The idiomatic way to deploy a workload in Kubernetes is that everything should be in a pod; however the current approach leaves files in the host filesystem such as CNI binaries and CNI configuration. These files are usually downloaded via the init container of the pod after binding, which increases the time for the pod to get to a running state. Since all existing K8s network plugins are running as daemonsets we will take this approach as well, where all the dependencies are packaged into the container image thus adopting a well known approach. This will have added benefits of the network pod startup being much faster as nothing should need to be downloaded. 

Another area that KNI will improve is ‘network readiness’. Currently the container runtime is involved with providing both network and runtime readiness with the Status CRI RPC. The container runtime defines the network as ‘ready’ by the presence of the CNI network configuration in the host file system. The more recent CNI specification does include a status verb, however this is still bound by the current limitations, files on disk and execution model. The KNI will provide an RPC that can be implemented so the kubelet will call the KNI via gRPC. 

KNI aims to help the community and other proposals in the Kubernetes ecosystem. We will do this by providing necessary information via the gRPC service. We should be the API that provides the “what networks are available on this node” so that another effort can make the kube-scheduler aware of networks. We should also provide IPAM status as a common issue, is that the IPAM runs out of assignable IP addresses and pods are no longer able to be scheduled on that node until intervention. We should provide visibility into this so that we can indicate “no more pods” as setting the node to not ready will evict the healthy pods. While the future state of KNI could aim to propose changes to the kube-scheduler, it's not a part of our initial work and instead should try to assist other efforts such as DRA/device plugin to provide the information they need. 

The community may ask for more features, as we are taking a bold approach to reimagining Kubernetes networking by reducing the amount of layers involved in networking. We should prioritize feature parity with the current CNI model and then capture future work. KNI aims to be the foundational network api that is specific for Kubernetes. 

```

Author: @shaneutt @danwinship this is my draft of the motivation.

Reviewer: you should merge it into the doc (with shorter lines) so we can comment on it

Reviewer: but in general, yes, it looks good

Author: @danwinship I copied both summary/motivation into the doc: https://docs.google.com/document/d/1Gz7iNtJNMI-zKJhaOcI3aflPCx3etJ01JMxzbtvruKk/edit?usp=sharing

- Guarantee the network is setup and in a healthy state before containers are started (ephemeral, init, regular)
- If feasible, provide API awareness of Pod network namespaces (e.g. interface names)
- Provide a uniform approach for network setup/teardown for both virtualized (kata) and non-virtualized (runc) 
runtimes including kubevirt. This could eliminate the high and low level runtimes from the networking path
- Provide a reference implementation of the KNI network runtime
- Provide the ability to have all the dependencies packaged in the container image (no more CNI binaries in the host file system)
..- No more downloading CNI binaries via initContainers/Mounting /etc/cni/net.d or /opt/cni/bin
- Provide the ability to use native k8s resources for configuration such as a ConfigMap's instead of configuration files in host file system
- Provide an API to indicate network readiness for the node (no more files on disk)
- Eliminate the need to exec binaries and replace with gRPC
- Make troubleshooting easier by having logs accessible via kubectl logs
- Improve network pod startup time

> - Provide the ability to prevent additional scheduling of pods if IPAM is out of IP addresses without evicting running pods

Author: I think we need to rephrase this one to be providing the API around the IPAM state so that we don't scope creep.

### Non-Goals

1. Any changes to the kube-scheduler 
2. Any specific implementation other than the reference implementation. However we should ensure the KNI-API is flexible enough to support

## Proposal

The proposal of this KEP is to design and implement the KNI-API and make necessary changes to the CRI-API and container runtimes. The scope should be kept to a minimum and we should target feature parity. 

### User Stories

We are constantly adding these user stories, please join the community sync to discuss. 

#### Story 1

> As a cluster operator, I need the ability to determine my network(s) is ready so that my pods come up with a working network.

Reviewer 4: this has to have more details about what mean "network is ready", there is no such a thing as a "global network state", the whole point of IP networks is to be distributed, 

"Network is ready" today means I can provide a netdevice to the Pod (veth) and assign an IP addresses

Author: Currently 'network is ready' is a CNI network configuration in /etc/cni/net.d. At a minimum, we should clearly define what 'network is ready' means. I can ask several people about this and get various answers depending on their environment. We should allow the user to implement this assuming it meets the criteria of K8s.

Reviewer 4: My view is that "network ready" today is expected to mean "I can create a Pod that will be able to communicate with all the other Pods in the cluster", and this is a paradox https://github.com/kubernetes/enhancements/pull/4477/files#r1489293806, however, it is implemented today as "there is a cni config file that we are expecting to do what is right when we create a pod", and this use to mean "I can create a Pod and wil be able to communicate within the node"

@danwinship @thockin and @squeed on these philosophical questions :)

#### Story 2

> As a cluster operator, I need the ability to determine what networks are available on my node so that upstream components can ensure the pod is scheduled on the appropriate node. 

Reviewer 4: I don't understand this user story, this is an scheduling problem that is already solved today https://kubernetes.io/docs/concepts/scheduling-eviction/topology-spread-constraints/

Author: I have been in debate about whether or not to focus more on interfaces and not networks, taking a more CNI approach. However, if one was to implement multiple networks most likely they would do it with different interfaces. Thanks for pointing out the topology spread constraints. In the end we don't want to be involved with scheduling however we want to provide the API to give other efforts what is currently on this node.

> aimed to keep this first iteration small and easier to consume. We are only
> focusing on the "What" and "Why" (e.g. motivation, goals, user stories) for
> this iteration so that we can build consensus on those first before we add
> any of the "How".

<!-- toc -->
- [Summary](#summary)
- [Motivation](#motivation)
  - [Goals](#goals)
  - [Non-Goals](#non-goals)
- [Proposal](#proposal)
  - [User Stories (Optional)](#user-stories-optional)
    - [Story 1](#story-1)
    - [Story 2](#story-2)
<!-- /toc -->

## Summary

KNI or Kubernetes Networking Interface, is an effort to take a second look at Kubernetes networking and evaluate what are the pain points, and what can we improve. At its core, KNI will be a foundational network API specific for Kubernetes that will provide flexibility and extensibility to solve basic and the most advanced and opinionated networking use cases.  
 
## Motivation

Kubernetes networking is an area of complexity and multiple layers which has created several challenges and areas of improvement. These challenges include deployment of the CNI plugins, troubleshooting networking issues and development of new functionality. 

> Currently networking happens in three layers of the stack, Kubernetes itself by means of kube-proxy or another controller based solution, the container runtime with network namespace creation and CNI plugins and the OCI runtime which does additional network setup for kernel isolated pods. All of this communication happens through non network specific APIs which to the reader of the code makes it hard to determine where ‘networking’ is happening. Having networking in several layers presents an issue when needing to troubleshoot issues as one needs to check several areas and some cannot be done via kubectl logs such as the CNI execution logs. This becomes more of an effort as multiple uncoordinated processes are making changes to the same resource, the network namespace of either the root or pod. The KNI aims at reducing the complexity by consolidating the networking into a single layer and having a uniform process for both namespaced and kernel isolated pods through a gRPC API. Leveraging gRPC will allow users the ability to migrate away from the current execution model that the CNI currently leverages.   

Reviewer 7: What is meant by `namespaced AND kernel-isolated pods`? isn't that the same thing?

Author: What I was trying to do here is separate the virtualized oci runtimes from the non-virtualized oci runtimes. Aka kata vs runc. However we just got off a call with Kubevirt where you could use either. Both will leverage network namespaces however the virtualized cases have the additional kernel isolation.

Reviewer 7: If both will use (at least) network namespacing it's probably less confusing to just say that (and provide more detail later in the doc for other cases if need be).

> Currently networking happens in three layers of the stack, Kubernetes itself by means of kube-proxy or another controller based solution, the container runtime with network namespace creation and CNI plugins and the OCI runtime which does additional network setup for kernel isolated pods. All of this communication happens through non network specific APIs which to the reader of the code makes it hard to determine where ‘networking’ is happening. Having networking in several layers presents an issue when needing to troubleshoot issues as one needs to check several areas and some cannot be done via kubectl logs such as the CNI execution logs. This becomes more of an effort as multiple uncoordinated processes are making changes to the same resource, the network namespace of either the root or pod. The KNI aims at reducing the complexity by consolidating the networking into a single layer and having a uniform process for both namespaced and kernel isolated pods through a gRPC API. Leveraging gRPC will allow users the ability to migrate away from the current execution model that the CNI currently leverages.   

Reviewer 8: Nothing in this document outlines how adding a gRPC API will reduce layers versus CNI.
Using gRPC will not inherently reduce layers versus RPC via exec.
How specifically does KNI intend to reduce layers?

> The next challenge is the deployment of the CNI plugins that provide the network setup and teardown of the pod. The idiomatic way to deploy a workload in Kubernetes is that everything should be in a pod; however the current approach leaves files in the host filesystem such as CNI binaries and CNI configuration. These files are usually downloaded via the init container of the pod after binding, which increases the time for the pod to get to a running state. Since all existing K8s network plugins are running as daemonsets we will take this approach as well, where all the dependencies are packaged into the container image thus adopting a well known approach. This will have added benefits of the network pod startup being much faster as nothing should need to be downloaded. 

Reviewer 7: A key part of what current CNI provides is a framework for different vendors to independently add extensions (CNI plugins) to a node, and have guarantees about when in the lifecycle of pod creation those are invoked, and what context they have from other plugins.

There may be multiple daemonsets from multiple vendors installing multiple CNI plugins on the same node, and any model we adopt must reckon with this as a first-class concern.

That, for me, is critical to retain for this to be a 1-1 replacement to the existing CNI - we can probably do something simpler than the current model of "kubelet talks to CRI, CRI impl reads shared node config with plugin list, serially runs the plugins as arbitrary binaries", as well.

I would want to see current CNI plugins from a few different projects implemented as KNI extensions, and installed/managed on the same node. If we can do that, we are effectively proving this can be a well-scoped replacement for the status quo.

> The next challenge is the deployment of the CNI plugins that provide the network setup and teardown of the pod. The idiomatic way to deploy a workload in Kubernetes is that everything should be in a pod; however the current approach leaves files in the host filesystem such as CNI binaries and CNI configuration. These files are usually downloaded via the init container of the pod after binding, which increases the time for the pod to get to a running state. Since all existing K8s network plugins are running as daemonsets we will take this approach as well, where all the dependencies are packaged into the container image thus adopting a well known approach. This will have added benefits of the network pod startup being much faster as nothing should need to be downloaded. 

Reviewer 7: >These files are usually downloaded via the init container of the pod after binding, which increases the time for the pod to get to a running state.

This is confusing me a bit - typically the CNI config is something workload pods are not aware of  at all, via `init` containers or otherwise - only the CRI implementation handles them?

Author: The pod I was referring to was the network plugin daemonset pods (flannel, calico, ...). I can try and clean this up to be more clear.

Reviewer 7: Ah ok thanks, yeah - "node agent daemonset" or "privileged node-bound pod" or whatever. Something general and consistent, since I don't think it _has_ to be an init container specifically, in the Pod sense.

Reviewer 9: Please see also https://github.com/kubernetes/enhancements/pull/4477#issuecomment-1953692060, and the followup comments

> The next challenge is the deployment of the CNI plugins that provide the network setup and teardown of the pod. The idiomatic way to deploy a workload in Kubernetes is that everything should be in a pod; however the current approach leaves files in the host filesystem such as CNI binaries and CNI configuration. These files are usually downloaded via the init container of the pod after binding, which increases the time for the pod to get to a running state. Since all existing K8s network plugins are running as daemonsets we will take this approach as well, where all the dependencies are packaged into the container image thus adopting a well known approach. This will have added benefits of the network pod startup being much faster as nothing should need to be downloaded. 

Reviewer 7: > CNI plugin binaries can be pre-installed to the host and run entirely on the host without a daemonset and there are major cluster operators working this way.

Yep, some environments bake the nodes and don't let you change them (and the corollary of that is you can't use other CNIs or meshes). Some do.

Either way, the current gate is "you need node privilege to install CNI plugins" - some envs let pods assume node privileges, some do not and prebake everything. That gate should be maintained for KNI, so either option works, depending on operator need.

> The next challenge is the deployment of the CNI plugins that provide the network setup and teardown of the pod. The idiomatic way to deploy a workload in Kubernetes is that everything should be in a pod; however the current approach leaves files in the host filesystem such as CNI binaries and CNI configuration. These files are usually downloaded via the init container of the pod after binding, which increases the time for the pod to get to a running state. Since all existing K8s network plugins are running as daemonsets we will take this approach as well, where all the dependencies are packaged into the container image thus adopting a well known approach. This will have added benefits of the network pod startup being much faster as nothing should need to be downloaded. 

Reviewer 8: > Either way, the current gate is "you need node privilege to install CNI plugins"  [...]

I would expect that to be true, given the position they fill in the stack (!)

> The next challenge is the deployment of the CNI plugins that provide the network setup and teardown of the pod. The idiomatic way to deploy a workload in Kubernetes is that everything should be in a pod; however the current approach leaves files in the host filesystem such as CNI binaries and CNI configuration. These files are usually downloaded via the init container of the pod after binding, which increases the time for the pod to get to a running state. Since all existing K8s network plugins are running as daemonsets we will take this approach as well, where all the dependencies are packaged into the container image thus adopting a well known approach. This will have added benefits of the network pod startup being much faster as nothing should need to be downloaded. 

Reviewer 8: > However, the piece that is important is that you no longer need to have any files in the host filesystem such as CNI binaries or CNI configuration files.

But _why_ is that important? I mean taken to an extreme ... You cannot run Kubernetes without files on the host filesystem anyhow, and configuring the pod network is an incredibly privileged place to operate.

> The next challenge is the deployment of the CNI plugins that provide the network setup and teardown of the pod. The idiomatic way to deploy a workload in Kubernetes is that everything should be in a pod; however the current approach leaves files in the host filesystem such as CNI binaries and CNI configuration. These files are usually downloaded via the init container of the pod after binding, which increases the time for the pod to get to a running state. Since all existing K8s network plugins are running as daemonsets we will take this approach as well, where all the dependencies are packaged into the container image thus adopting a well known approach. This will have added benefits of the network pod startup being much faster as nothing should need to be downloaded. 

Reviewer 7: > But why is that important? I mean taken to an extreme ... You cannot run Kubernetes without files on the host filesystem anyhow, and configuring the pod network is an incredibly privileged place to operate.

It's important because allowing privileged pods to run on a node (where privileged necessarily means "can mutate node state") is an operator choice today, and it seems wrong to take that choice away from operators.

This is how, for instance, it is possible to install Cilium in AWS today. Or Calico in GKE. Or Openshift, etc etc.

Anyone that doesn't want to allow privileged pods on a node can leverage existing K8S APIs to preclude that as a matter of operational policy - it's rather orthogonal to CNI.

Another area that KNI will improve is ‘network readiness’. Currently the container runtime is involved with providing both network and runtime readiness with the Status CRI RPC. The container runtime defines the network as ‘ready’ by the presence of the CNI network configuration in the host file system. The more recent CNI specification does include a status verb, however this is still bound by the current limitations, files on disk and execution model. The KNI will provide an RPC that can be implemented so the kubelet will call the KNI via gRPC. 

> KNI aims to help the community and other proposals in the Kubernetes ecosystem. We will do this by providing necessary information via the gRPC service. We should be the API that provides the “what networks are available on this node” so that another effort can make the kube-scheduler aware of networks. We should also provide IPAM status as a common issue, is that the IPAM runs out of assignable IP addresses and pods are no longer able to be scheduled on that node until intervention. We should provide visibility into this so that we can indicate “no more pods” as setting the node to not ready will evict the healthy pods. While the future state of KNI could aim to propose changes to the kube-scheduler, it's not a part of our initial work and instead should try to assist other efforts such as DRA/device plugin to provide the information they need. 

Reviewer 8: > We should also provide IPAM status as a common issue, is that the IPAM runs out of assignable IP addresses and pods are no longer able to be scheduled on that node until intervention.

Typically this is handled by coordinating the assigned IP range and the max pods setting in the kubelet.
Cluster operators already have the tools to prevent this issue, why would you allow a node to be configured to have more pods than IPs?

> The community may ask for more features, as we are taking a bold approach to reimagining Kubernetes networking by reducing the amount of layers involved in networking. We should prioritize feature parity with the current CNI model and then capture future work. KNI aims to be the foundational network api that is specific for Kubernetes and should make troubleshooting easier, deploying more friendly and innovate faster while reducing the need to make changes to core Kubernetes. 

Reviewer 8: > The community may ask for more features, as we are taking a bold approach to reimagining Kubernetes networking by reducing the amount of layers involved in networking.

Reducing the layers ... by adding a new GRPC service? Why isn't this CRI? (An existing Kubernetes-Specific gRPC service for Pods, which includes some networking related bits today ...)

Author: Adding this to the CRI-API is a design detail. The only piece of information relevant to networking is the PodIP coming back in PodSandboxStatus. This talks about eliminating the network setup/teardown and netns creation from the container/oci runtimes.

Reviewer 8: > This talks about eliminating the network setup/teardown and netns creation from the container/oci runtimes.

Has this been a significant obstacle for implementers? Examples?

What blocks a CNI revision from handling netns/teardown?

### Goals

- Design a cool looking t-shirt

> - Provide a RPC for the Attachment and Detachment of interface[s] for a Pod

Reviewer 8: This seems like a significant departure from current expectations around pod networking.

Author: Can you clarify this? We already do something similar with CNI ADD/DEL with an execution model. This is levering gRPC to communicate with the gRPC server that would be flannel, calico, cilium.

- Provide a RPC for the Querying of Pod network information (interfaces, network namespace path, ip addresses, routes, ...)

> - Provide a RPC to prevent additional scheduling of pods if IPAM is out of IP addresses without evicting running pods

Reviewer 8: Surely this is just reporting status up through CRI?

Author: Last place I mention this, that the decision around CRI or a new API is a design detail.

- Provide a RPC to indicate network readiness for the node (no more CNI network configuration files in host file system)
- Provide a RPC to provide the user the ability to query what networks are on a node
- Consolidate K8s networking to a single layer without involving the container/oci runtimes
- KNI should provide the RPC's required to establish feature parity with current CNI [ADD, DEL]
- Provide documentation, examples, troubleshooting and FAQ's for KNI.
- Decouple the Pod and Node Network setup
- Provide garbage collection to ensure no resources created during pod setup such as Linux bridges, ebpf programs, 
allocated IP addresses are left behind after pod deletion

> - Improve the current IP handling for pods (PodIP) to be handle multiple IP addresses and 

Reviewer 8: Which is not possible in CNI because ...?

a field to identify the IP address family (IPV4 vs IPV6)
- Provide backwards compatibility for the existing CNI approach and migration a path to fully adopt KNI
- Provide a uniform approach for network setup/teardown for both virtualized (kata) and non-virtualized (runc) 
runtimes including kubevirt. This could eliminate the high and low level runtimes from the networking path
- Provide a reference implementation of the KNI network runtime
- Provide the ability to have all the dependencies packaged in the container image (no more CNI binaries in the host file system)

> ..- No more downloading CNI binaries via initContainers/Mounting /etc/cni/net.d or /opt/cni/bin

Reviewer 8: ... Downloading a container image to run the gRPC service is better how?

Currently a cluster operator can just pre-populate these instead of relying on a daemonset, it's pretty straightforward and results in very fast startup.

Reviewer 7: That's valid, I think - there needs to be a story for supporting pre-baked nodes.

That could be as simple as shipping the node with KNI as a system service, or shipping a node with a preloaded image.

Author: This doesn't make too much sense. Since calico, flannel, cilium are already running as daemonset pods. Since they would be implementing the KNI grpc service. Are you looking for a migration path here?

Reviewer 7: There's 2 (somewhat edge case but still valid) operator questions that can be answered here.

1. I am an operator and I want to ship nodes with a specific CNI impl and not allow anyone downstream to change that, how?

2. I am an operator, and I want to set up airgapped (or at least, "fully preloaded") clusters that don't require dynamically fetching container image bytes to become functional for workload deployment.

Both of these can be solved with existing K8S primitives and patterns, with or without KNI, so I don't think KNI makes this materially more difficult.

- Provide the ability to use native k8s resources for configuration such as a ConfigMap's instead of configuration files in host file system
- Eliminate the need to exec binaries and replace with gRPC
- Make troubleshooting easier by having network runtime logs accessible via kubectl logs
- Improve network pod startup time

### Non-Goals

1. Any changes to the kube-scheduler 
2. Any specific implementation other than the reference implementation. However we should ensure the KNI-API is flexible enough to support

## Proposal

The proposal of this KEP is to design and implement the KNI-API and make necessary changes to the CRI-API and container runtimes. The scope should be kept to a minimum and we should target feature parity. 

### User Stories

> We are constantly adding these user stories, please join the community sync to discuss. 

Reviewer 8: > We are constantly adding these user stories [...]

Where?

### Notes/Constraints/Caveats

## Constraints

1. Guarantee the pod interface is setup and in a healthy state before containers are started (ephemeral, init, regular)

## Notes

> Additional Information/Diagrams: https://docs.google.com/document/d/1Gz7iNtJNMI-zKJhaOcI3aflPCx3etJ01JMxzbtvruKk/edit?usp=sharing

Reviewer 8: Anything relevant to the KEP should be in-lined and committed to the repo.

Changes to the pod specification will require hard evidence. 

The specifics of "Network Readiness" is an implementation detail. We need to provide this RPC to the user. 

Since the network runtime can be run separated from the container runtime, you can package everything into a pod and not need to have binaries on disk. This allows the CNI plugins to be isolated in the pod and the pod will never need to mount /opt/cni/bin or /etc/cni/net.d. This offers a potentially more ability to control execution. Keep in mind CNI is the implementation however when this is used chaining is still available.

> ## Ongoing Considerations

Reviewer 8: There doesn't seem to be any discussion of how we might improve CNI, CRI instead and why that isn't sufficient versus this entirely new API and RPC service.

Author: This could live in the CRI-API as multiple services, no one has indicated that we must use a new API. However the CNI 2.0 was talked about being closer to K8s for years now. This is that effort.

Reviewer 7: I do agree we could probably  be more clear up front in the KEP about why the current CNI model (slurping out-of-k8s configs from well-known paths, TOCTOU errors, telling CRI implementations via out-of-band mechanisms to exec random binaries on the node) is somewhat inefficient, error-prone, operationally fragile, and not strictly necessary.

# KEP-4410: Kubernetes Networking reImagined

> **NOTE**: for the initial PR we've removed a lot of the templated text and
> aimed to keep this first iteration small and easier to consume. We are only
> focusing on the "What" and "Why" (e.g. motivation, goals, user stories) for
> this iteration so that we can build consensus on those first before we add
> any of the "How".

<!-- toc -->
- [Summary](#summary)
- [Motivation](#motivation)
  - [Goals](#goals)
  - [Non-Goals](#non-goals)
- [Proposal](#proposal)
  - [User Stories (Optional)](#user-stories-optional)
    - [Story 1](#story-1)
    - [Story 2](#story-2)
<!-- /toc -->

## Summary

KNI or Kubernetes Networking Interface, is an effort to take a second look at Kubernetes networking and evaluate what are the pain points, and what can we improve. At its core, KNI will be a foundational network API specific for Kubernetes that will provide flexibility and extensibility to solve basic and the most advanced and opinionated networking use cases.  
 
## Motivation

> Kubernetes networking is an area of complexity and multiple layers which has created several challenges and areas of improvement. These challenges include deployment of the CNI plugins, troubleshooting networking issues and development of new functionality. 

Reviewer 8: Yes, please. The current formatting has too many discussion points per line, it's difficult to follow and comment on.
